extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java

```diff
@@ -0,0 +1,747 @@
+    try {
+      catalog.validate(table);
+    }
+    catch (IAE e) {
+      return Actions.badRequest(Actions.INVALID, e.getMessage());
+    }
+
+    switch (action) {
+      case NEW:
+        return insertTableSpec(table, false);
+      case IFNEW:
+        return insertTableSpec(table, true);
+      case REPLACE:
+        return updateTableSpec(table, version);
+      case FORCE:
+        return addOrUpdateTableSpec(table);
+      default:
+        throw new ISE("Unknown action.");
+    }
+  }
+
+  private Response authorizeTable(TableId tableId, TableSpec spec, final HttpServletRequest req)
+  {
+    // Druid has a fixed set of schemas. Ensure the one provided is valid.
+    Pair<Response, SchemaSpec> result = validateSchema(tableId.schema());
+    if (result.lhs != null) {
+      return result.lhs;
+    }
+    SchemaSpec schema = result.rhs;
+
+    // The schema has to be one that allows table definitions.
+    if (!schema.writable()) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format("Cannot modify schema %s", tableId.schema())
+      );
+    }
+
+    // Table name can't be blank or have spaces
+    if (Strings.isNullOrEmpty(tableId.name())) {
+      return Actions.badRequest(Actions.INVALID, "Table name is required");
+    }
+    if (!tableId.name().equals(tableId.name().trim())) {
+      return Actions.badRequest(Actions.INVALID, "Table name cannot start or end with spaces");
+    }
+
+    // The user has to have permission to modify the table.
+    try {
+      catalog.authorizer().authorizeTable(schema, tableId.name(), Action.WRITE, req);
+    }
+    catch (ForbiddenException e) {
+      return Actions.forbidden(e);
+    }
+
+    // Validate the spec, if provided.
+    if (spec != null) {
+
+      // The given table spec has to be valid for the given schema.
+      if (Strings.isNullOrEmpty(spec.type())) {
+        return Actions.badRequest(Actions.INVALID, "Table type is required");
+      }
+
+      if (!schema.accepts(spec.type())) {
+        return Actions.badRequest(
+            Actions.INVALID,
+            StringUtils.format(
+                "Cannot create tables of type %s in schema %s",
+                spec.getClass().getSimpleName(),
+                tableId.schema()
+            )
+        );
+      }
+    }
+
+    // Everything checks out, let the request proceed.
+    return null;
+  }
+
+  private Response insertTableSpec(TableMetadata table, boolean ifNew)
+  {
+    try {
+      long createVersion = catalog.tables().create(table);
+      return Actions.okWithVersion(createVersion);
+    }
+    catch (DuplicateKeyException e) {
+      if (!ifNew) {
+        return Actions.badRequest(
+              Actions.DUPLICATE_ERROR,
+              StringUtils.format(
+                  "A table of name %s already exists",
+                  table.id().sqlName()
+              )
+        );
+      } else {
+        return Actions.okWithVersion(0);
```

Comment:
Seems weird to return a version of 0 here, why not return the existing version? Maybe the create method called above should have a version that takes `ifNew` and returns the current version if the table already exists?

Reply:
In response to the potential for information leak, wouldnt the user already be authorized before getting far enough to get a duplicate error? If not they should be, in which case I think we would only leak the version information to an authorized user.